pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,648 @@
+	// are scaled up.
+	for _, pod := range pods {
+		for _, ip := range pod.ips {
+			for _, portProtocol := range config.backendPorts {
+				var portInt uint16
+				if iana.IsSvcName(portProtocol.Port) {
+					if port, ok := pod.namedPorts[portProtocol.Port]; ok {
+						portInt = port.Port
+					}
+				} else {
+					port, _ := strconv.ParseUint(portProtocol.Port, 0, 16)
+					portInt = uint16(port)
+				}
+				beIP := net.ParseIP(ip)
+				if beIP == nil {
+					continue
+				}
+				be := backend{
+					IP: net.ParseIP(ip),
+					L4Addr: lb.L4Addr{
+						Protocol: lb.L4Type(portProtocol.Protocol),
+						Port:     portInt,
+					},
+				}
+				if config.frontendMappingForAddr.feAddr.IP.To4() != nil {
+					if option.Config.EnableIPv4 {
+						bes4 = append(bes4, be)
+					}
+				} else {
+					if option.Config.EnableIPv6 {
+						bes6 = append(bes6, be)
+					}
+				}
+			}
+
+		}
+		if len(bes4) > 0 {
+			rpm.upsertServiceWithBackends(config, config.frontendMappingForAddr, pod.id, bes4)
+		} else if len(bes6) > 0 {
+			rpm.upsertServiceWithBackends(config, config.frontendMappingForAddr, pod.id, bes6)
+		}
+	}
+	return
+}
+
+// upsertForConfigTypeSvc upserts service(s) for the given policy config service frontend(s).
+func (rpm *Manager) upsertForConfigTypeSvc(config *LRPConfig, pods ...*podMetadata) {
+	// Generate backends for the policy config's backend named ports, and then
+	// map the backends to policy frontends based on the named ports.
+	// We currently don't check which backends are updated before upserting a
+	// a service with the corresponding frontend. This can be optimized if LRPs
+	// are scaled up.
+	for _, bePort := range config.backendPorts {
+		namedPort := bePort.Port
+		var bes4 []backend
+		var bes6 []backend
+		if feM, ok := config.frontendMappingForSvc[namedPort]; ok {
+			for _, pod := range pods {
+				if port, ok := pod.namedPorts[namedPort]; ok {
+					// Generate pod backends.
+					for _, ip := range pod.ips {
+						beIP := net.ParseIP(ip)
+						if beIP == nil {
+							continue
+						}
+						be := backend{
+							IP: net.ParseIP(ip),
+							L4Addr: lb.L4Addr{
+								Protocol: port.Protocol,
+								Port:     port.Port,
+							},
+						}
+						if feM.feAddr.IP.To4() != nil {
+							if option.Config.EnableIPv4 {
+								bes4 = append(bes4, be)
+							}
+						} else {
+							if option.Config.EnableIPv6 {
+								bes6 = append(bes6, be)
+							}
+						}
+					}
+				}
+				if len(bes4) > 0 {
+					rpm.upsertServiceWithBackends(config, feM, pod.id, bes4)
+				} else if len(bes6) > 0 {
+					rpm.upsertServiceWithBackends(config, feM, pod.id, bes6)
+				}
+			}
+		}
+	}
+}
+
+// upsertServiceWithBackends updates policy config internal state and upserts
+// service with the given pod backends.
+func (rpm *Manager) upsertServiceWithBackends(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
+	frontendMapping.backends = backends
```

Comment:
`rpm.upsertService()` is silently consuming any error in attempting to upsert the service a few lines below, so in the event that there's a problem inside that function callstack, this path will have already updated `frontendMapping.backends` so from the redirectmanager's understanding it will think that these backends are plumbed, but in practice they will not be. I think it would be more robust to handle the error up the stack and only make modifications like this if the upsert is successful?

Reply:
^ The above assumes this field represents "realized state", in which case you want the cache to reflect the actual state in the datapath. If it's in fact desired state then the error handling would be different.